### PR TITLE
client: tests for getTrieNodes and trieNodes SNAP messages

### PR DESCRIPTION
**Pull request by scorbajio**
This PR adds tests for the getTrieNodes and trieNodes SNAP messages. Real data is captured and used for tests.

This PR also adds support for compact encoding for use with `getTrieNodes` and `trieNodes`.

### CODE REVIEW

**Comment by holgerd77** [comment]
Just to note: CI tests are partly failing for this PR.

**Comment by scorbajio** [comment]
> 

Seems like they

> Just to note: CI tests are partly failing for this PR.

I made some linting fixes but it still seems to be failing due to a download error. Is there

**Comment by scorbajio** [comment]
> 

I am not sure what to make of the karma browser test failures. It seems to be complaining about a dist file. @acolytec3 are these karma errors a known issue?

**Comment by acolytec3** [comment]
> > 
> 
> I am not sure what to make of the karma browser test failures. It seems to be complaining about a dist file. @acolytec3 are these karma errors a known issue?

I'll take a look in the morning. Karma can be tricky

**Comment by acolytec3** [comment]
> > > 
> > 
> > 
> > I am not sure what to make of the karma browser test failures. It seems to be complaining about a dist file. @acolytec3 are these karma errors a known issue?
> 
> I'll take a look in the morning. Karma can be tricky

Fixed.  We had a missing configuration option for karma.  Tests are passing locally now.

File: packages/client/test/net/protocol/snapprotocol.spec.ts [hunk 1]
```diff
@@ -456,6 +456,86 @@ tape('[SnapProtocol]', (t) => {
...
+
+    const reqId = BigInt(1)
+    const root = Buffer.from(
+      '04157502e6177a76ca4dbf7784e5ec1a926049db6a91e13efb70a095a72a45d9',
+      'hex'
+    )
+    const paths = [[Buffer.from('0x0', 'hex')], [Buffer.from('0x0', 'hex')]]
+    const bytes = BigInt(5000000)
+
+    const payload = p.encode(p.messages.filter((message) => message.name === 'GetTrieNodes')[0], {
```
Comment (jochem-brouwer) [inline packages/client/test/net/protocol/snapprotocol.spec.ts:473]: I'm not entire sure what this line does (looks a bit like magic to me), it filters out the right message name and then assigns the values to it..? How does this magically call the `encode` then..?

File: packages/util/src/encoding.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+export const hasTerm = (s: Uint8Array) => {
```
[the remark this answers is not in the document]
Comment (scorbajio) [reply packages/util/src/encoding.ts:1]: I haven't been able to find a reason for why 16 is used as the terminator, so I assume it is some unspoken convention. [This](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/) article also describes the terminator as optional.

**Comment by scorbajio** [comment]
> Great work! I have some questions. Do you also have an URL to this compact encoding somewhere?

I wasn't able to find much on compact encoding. There is some mention of it [here](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/), and the code was ported from [here](https://github.com/ethereum/go-ethereum/blob/master/trie/encoding.go).

File: packages/util/src/encoding.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (scorbajio) [reply packages/util/src/encoding.ts:6]: I added docs for the method, but I haven't been able to find a reason for why 16 is used as the terminator, so I assume it is some unspoken convention. [This](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/) article also describes the terminator as optional.

**Comment by scorbajio** [comment]
> Great work! I have some questions. Do you also have an URL to this compact encoding somewhere?

I wasn't able to find much on compact encoding. There is some mention of it [here](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/), and the code was ported from [geth's trie/encoding.go](https://github.com/ethereum/go-ethereum/blob/master/trie/encoding.go).

File: packages/client/test/net/protocol/snapprotocol.spec.ts [hunk 3]
```diff
@@ -504,3 +584,17 @@ const getByteCodesRLP =
...
+//   paths: [
+//     [Buffer.from("0x0", "hex")],
+//     [Buffer.from("0x0", "hex")],
+//   ],
+//   bytes: BigInt(5000000),
+// })
+// TODO getTrieNodesRLP is currently unused
+const _getTrieNodesRLP =
+  'eb01a004157502e6177a76ca4dbf7784e5ec1a926049db6a91e13efb70a095a72a45d9c4c100c100834c4b40'
+const trieNodesRLP =
```
[the remark this answers is not in the document]
Comment (scorbajio) [reply packages/client/test/net/protocol/snapprotocol.spec.ts:599]: I added the number and network in a comment right above, but I am not sure what you mean by request. If you mean the request payload, it is the value of `_getTrieNodesRLP`.

File: packages/client/test/net/protocol/snapprotocol.spec.ts [hunk 4]
```diff
@@ -456,6 +456,86 @@ tape('[SnapProtocol]', (t) => {
...
+    const config = new Config({ transports: [] })
+    const chain = new Chain({ config })
+    const p = new SnapProtocol({ config, chain })
+
+    const reqId = BigInt(1)
+    const root = Buffer.from(
+      '04157502e6177a76ca4dbf7784e5ec1a926049db6a91e13efb70a095a72a45d9',
+      'hex'
+    )
+    const paths = [[Buffer.from('0x0', 'hex')], [Buffer.from('0x0', 'hex')]]
```
[the remark this answers is not in the document]
Comment (scorbajio) [reply packages/client/test/net/protocol/snapprotocol.spec.ts:470]: Fixed.

**Review by g11tech — APPROVED** [review]
lgtm!

**Pull request merged**